### PR TITLE
pastebinit: fix deprecation warning and add darwin compatibility

### PR DESCRIPTION
**Pull request by Samuel-Martineau**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/pastebinit/default.nix [hunk 1]
```diff
@@ -21,11 +21,9 @@ stdenv.mkDerivation rec {
...
   patches = [
     # Required to allow pastebinit 1.5 to run on Python 3.8
-    (fetchpatch {
-      name = "use-distro-module.patch";
-      url = "https://bazaar.launchpad.net/~arnouten/pastebinit/python38/diff/264?context=3";
-      sha256 = "1gp5inp4xald65xbb7fc5aqq5s2fhw464niwjjja9anqyp3zhawj";
-    })
+    ./use-distro-module.patch
+    # Required to remove the deprecation warning of FancyURLopener
+    ./use-urllib-request.patch
```
Comment (raboof) [inline pkgs/tools/misc/pastebinit/default.nix:26]: have you also submitted this patch upstream?

**Comment by raboof** [comment]
Result of `nixpkgs-review pr 268420` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pastebinit</li>
  </ul>
</details>

File: pkgs/tools/misc/pastebinit/default.nix [hunk 2]
```diff
@@ -47,6 +45,6 @@ stdenv.mkDerivation rec {
     description = "A software that lets you send anything you want directly to a pastebin from the command line";
     maintainers = with maintainers; [ raboof ];
     license = licenses.gpl2;
-    platforms = platforms.linux;
+    platforms = platforms.linux ++ lib.platforms.darwin;
```
Comment (raboof) [inline pkgs/tools/misc/pastebinit/default.nix:48]: (I can't confirm darwin compatibility but will take your word for it ;) )

File: pkgs/tools/misc/pastebinit/default.nix [hunk 3]
```diff
@@ -21,11 +21,9 @@ stdenv.mkDerivation rec {
 
   patches = [
     # Required to allow pastebinit 1.5 to run on Python 3.8
-    (fetchpatch {
-      name = "use-distro-module.patch";
-      url = "https://bazaar.launchpad.net/~arnouten/pastebinit/python38/diff/264?context=3";
-      sha256 = "1gp5inp4xald65xbb7fc5aqq5s2fhw464niwjjja9anqyp3zhawj";
-    })
+    ./use-distro-module.patch
```
Comment (raboof) [inline pkgs/tools/misc/pastebinit/default.nix:24]: I see this patch needed to be updated, thanks for doing that. Have you also submitted this improvement upstream?

**Review by raboof — APPROVED** [review]
I think we should ideally submit these changes upstream and refer to those upstream patches in the nix packaging, instead of having a purely 'local' patch.

Otherwise LGTM.

**Comment by Samuel-Martineau** [comment]
> I think we should ideally submit these changes upstream and refer to those upstream patches in the nix packaging, instead of having a purely 'local' patch.
> 
> Otherwise LGTM.

I would tend to agree with you, but I couldn't find a patch for these upstream. The repo for `pastebinit` itself has not been updated since 2016 and most of the bug reports remain unanswered. However, I just realized that [Ubuntu maintains it's own version of `pastebinit`](https://launchpad.net/ubuntu/+source/pastebinit) (instead of https://launchpad.net/pastebinit/).

Do you think we should use the Ubuntu version as the new upstream?

File: pkgs/tools/misc/pastebinit/default.nix (hunk 2 above)
Comment (Samuel-Martineau) [reply pkgs/tools/misc/pastebinit/default.nix:48]: It runs on my Mac without any deprecation warnings :)

**Comment by raboof** [comment]
> > I think we should ideally submit these changes upstream and refer to those upstream patches in the nix packaging, instead of having a purely 'local' patch.
> > Otherwise LGTM.
> 
> I would tend to agree with you, but I couldn't find a patch for these upstream. The repo for `pastebinit` itself has not been updated since 2016 and most of the bug reports remain unanswered. However, I just realized that [Ubuntu maintains it's own version of `pastebinit`](https://launchpad.net/ubuntu/+source/pastebinit) (instead of https://launchpad.net/pastebinit/).
> 
> Do you think we should use the Ubuntu version as the new upstream?

Ah, that's a great observation. At first I thought it'd only be the packaging but indeed it also appears to include the script, and have updates not found in https://launchpad.net/pastebinit/. Let's indeed switch to that one.

With that in mind I think this is good to merge. Do you want to work on switching to the newer upstream as well?

**Comment by Samuel-Martineau** [comment]
@raboof I just submitted #268803 which moves to the maintained upstream and adds tests.

**Pull request merged**